Kernel/Bus/USB/Drivers/USBDriver.h

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2022, Jesse Buhagiar <jesse.buhagiar@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Error.h>
+#include <AK/NonnullRefPtr.h>
+#include <Kernel/Bus/USB/USBDevice.h>
+
+namespace Kernel::USB {
+
+typedef void (*driver_init_function_t)();
+#define USB_DEVICE_DRIVER(driver_name) \
+    driver_init_function_t driver_init_function_ptr_##driver_name __attribute__((section(".driver_init"), used)) = &driver_name::init
+
+class Driver : public AtomicRefCounted<Driver> {
+public:
+    Driver(StringView name)
+        : m_driver_name(name)
+    {
+    }
+
+    virtual ~Driver() = default;
+
+    virtual bool may_accept_device(USBDeviceDescriptor const& descriptor) = 0;
+    virtual ErrorOr<void> probe(USBInterface const& interface) = 0;
+    virtual ErrorOr<size_t> read(size_t count, u8* data) = 0;
+    virtual ErrorOr<size_t> write(size_t count, u8 const* data) = 0;
+    virtual void disconnect() = 0;
+    virtual StringView name() const { return m_driver_name; }
```

Comment:
Why is this virtual? The derived class should just be able to pass the right value to `Driver`'s constructor.

Reply:
Ah sure, missed that
(Previously it was pure virtual for some reason)